Fix inconsistent dependency to project in QgsRelation/QgsRelationManager

dmarteau opened the pull request:
## Description

Fix inconsistent dependency to project from QgsRelation/QgsRelationManager
by forcing  `QgsRelation::createFromXml` to rely on project passed
to `QgsRelationManager` instead of using `QgsProject::instance()`

This was leading to code like:
```
project = QgsProject()
project.read("/my/project/path.qgs")
rm = project.relationManager()
```
To return manager with no relations even if the project have ones, because relations are created from the the 'QgsProject::instance()' which is a different (potentially empty) project.

This PR ensure that: 

1. Updating relation status from manager that refer to a project then updates all relations with the same project.
2. Loading a project with relations leads to valid relations. 

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] Unit tests have been modified  for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [x] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

m-kuhn commented on src/core/qgsrelationmanager.cpp:
```diff
@@ -21,16 +21,13 @@
 
-QgsRelationManager::QgsRelationManager( QgsProject *project )
-  : QObject( project )
+QgsRelationManager::QgsRelationManager( QgsProject &project )
```
In general QObject based members are passed by pointer (not reference), just like Qt does too.

N.b. it would be nice if the server would follow this approach too.

dmarteau replied:
Hi @m-kuhn 
No problem to sacrifice to Qt standards but I'm always uncomfortable with the fact that you may pass null pointer when non null pointer is required and appreciate static typing for this - this why we use reference instead.

By the way, this is qgis core changes not only server.

m-kuhn replied:
Do a `Q_ASSERT( pointer )` and document it in the API to make your uncomfortability explicit.

dmarteau replied:
@m-kuhn  note that this is a pointer thas is passed to QObject - not the reference.

dmarteau replied:
Reverted signature in order to prevent API break. Still allow nullptr to preserve previous behavior.

elpaso commented on python/core/auto_generated/qgsrelation.sip.in:
```diff
@@ -45,10 +44,11 @@ changed.
 
 
-    static QgsRelation createFromXml( const QDomNode &node, QgsReadWriteContext &context );
+    static QgsRelation createFromXml( const QgsProject &project, const QDomNode &node, QgsReadWriteContext &context );
```
I'm afraid this qualifies as an API break, you should deprecate the old methods and add your new API.

m-kuhn commented on src/core/qgsrelation.h:
```diff
@@ -113,12 +113,13 @@ class CORE_EXPORT QgsRelation
      * \returns A relation
      */
-    static QgsRelation createFromXml( const QDomNode &node, QgsReadWriteContext &context );
+    static QgsRelation createFromXml( const QgsProject &project, const QDomNode &node, QgsReadWriteContext &context );
```
This is an api break. Either create a new overload or add it as an optional parameter (which can be turned into a mandatory parameter in QGIS 4)

dmarteau replied:
Ok, I'm changing this to an optional pointer and fallback fetching project fron `QgsProject::instance()` when passed pointer is null.

m-kuhn reviewed (commented):
Thanks, this is much desired.
Just two small things to adjust to make this mergable.

elpaso commented on src/core/qgsproject.cpp:
```diff
@@ -358,7 +358,7 @@ QgsProject::QgsProject( QObject *parent )
   , mBadLayerHandler( new QgsProjectBadLayerHandler() )
   , mSnappingConfig( this )
-  , mRelationManager( new QgsRelationManager( this ) )
+  , mRelationManager( new QgsRelationManager( *this ) )
```
I'd stick to the other managers convention and pass a pointer. We should probably move all of them to unique ptrs eventually but for now I'd recommend to be consistent with the other calls..

elpaso commented on src/core/qgsrelationmanager.h:
```diff
@@ -41,7 +41,7 @@ class CORE_EXPORT QgsRelationManager : public QObject
      * \param project associated project (used to notify project of changes)
      */
-    explicit QgsRelationManager( QgsProject *project = nullptr );
+    explicit QgsRelationManager( QgsProject &project );
```
Same considerations about API break and passing a pointer applies here.

dmarteau replied:
New commit that preserve API.

elpaso reviewed (commented):
Changes look good to me, I just left a few comments to be addressed.

rldhont commented:
@elpaso @m-kuhn do you think we can backport these changes even in release-3_4 ?

m-kuhn commented:
Thanks @dmarteau 

@rldhont does it fix a critical issue?

dmarteau commented:
> @rldhont does it fix a critical issue?

It prevents using relations from server plugin, service and more generally from pyQgis

m-kuhn commented:
I am afraid that it's too late in the development cycle to add this to 3.4. No objections to backport to 3.10 though.

rldhont commented:
@m-kuhn In the context of QGIS Server, this is a blocked issue for LTR users. When the project is parsed by QGIS Server, it does not contain relations so relations can't be used in this context for example in atlasprint https://github.com/3liz/qgis-atlasprint/issues/36

The workaround is to parsed QGIS project at each request, but we have build QGIS Server 3 to avoid it. Even if 3.10 can be fixed and will be the next LTR, LTR users will not upgrade to 3.10 before it becomes LTR. They will therefore not benefit from the patch for 3 months.

I would like to be able to correct version 3.4 until version 3.4.16

m-kuhn commented:
I very well understand.

At the same time, users have been able to live without this functionality for a year already. Are you able to get away with the workaround you describe as a band aid and convince them to update in a timely manner.
*Reasoning for being that cautious:* anything that's merged now must be absolutely 110% bulletproof, because we are very close to the last release in the 3.4 series. Forever. And we don't want a last 3.4 release that introduced regressions.

dmarteau commented:
ok, this is more complicated than expected, because QgsRelation was expecting working on QgsProject::instance at each step, I'm going to restore original tests to make sure that no regression occurs with the expected current behavior and add new specific tests to check that the fix do the job.

nyalldawson commented:
Could you make that a new "QgsRelationContext" class instead of just the project pointer? It makes it much easier to extend in future without breaking API.

dmarteau committed:
Pass project when creating relation from Xml

    Fix dependency inconsistency to project from QgsRelationManager
    by forcing  'QgsRelation::createFromXml' to rely on project passed
    to 'QgsRelationManager' instead of using 'QgisProject::instance()'

dmarteau commented:
@nyalldawson Yes,  good idea. I'm going to explore that path.

dmarteau commented:
@m-kuhn @elpaso I have refactored the whole thing according to @nyalldawson suggestion

m-kuhn commented on python/core/auto_generated/qgsrelation.sip.in:
```diff
@@ -33,10 +32,16 @@ class QgsRelation
     QgsRelation();
 %Docstring
-Default constructor. Creates an invalid relation.
+Default constructor with option param. Creates an invalid relation.
```
What does that mean exactly?

dmarteau replied:
just a typo from previous comment.

m-kuhn commented on python/core/auto_generated/qgsrelation.sip.in:
```diff
@@ -308,6 +314,8 @@ Gets the referencing field counterpart given a referenced field.
 Updates the validity status of this relation.
 Will be called internally whenever a member is changed.
 
+:param project: te get the layers from, default to global project instance.
```
Is this a leftover (and if no, I think there is a typo ;) )

m-kuhn commented on src/core/qgsrelation.cpp:
```diff
@@ -27,20 +27,28 @@ QgsRelation::QgsRelation()
 }
 
-QgsRelation QgsRelation::createFromXml( const QDomNode &node, QgsReadWriteContext &context )
+QgsRelation QgsRelation::createFromXml( const QDomNode &node, QgsReadWriteContext &context,  const QgsRelationContext &relcontext )
```
```suggestion
QgsRelation QgsRelation::createFromXml( const QDomNode &node, QgsReadWriteContext &context,  const QgsRelationContext &relationContext )
```

m-kuhn commented on src/core/qgsrelation.h:
```diff
@@ -32,7 +33,7 @@ class QgsFeatureRequest;
 class QgsVectorLayer;
 class QgsRelationPrivate;
-
+class QgsProject;
```
```suggestion
```

m-kuhn commented on src/core/qgsrelation.h:
```diff
@@ -91,11 +92,17 @@ class CORE_EXPORT QgsRelation
 
+    /**
+     * Constructor with context. Creates an invalid relation.
+     * /param context a relation context
```
```suggestion
```

No reason to state the obvious :)

m-kuhn commented on src/core/qgsrelation.h:
```diff
@@ -371,13 +379,17 @@ class CORE_EXPORT QgsRelation
      * Updates the validity status of this relation.
      * Will be called internally whenever a member is changed.
      *
+     * \param project te get the layers from, default to global project instance.
```
Leftover?

m-kuhn commented on src/core/qgsrelationcontext.h:
```diff
@@ -0,0 +1,66 @@
+/**
+ * \ingroup core
+ * \class QgsRelationContext
+ * Associate the current context to the relations
```
```suggestion
 * Context for relations. Used to resolve layers from projects.
```

m-kuhn commented on src/core/qgsrelationcontext.h:
```diff
@@ -0,0 +1,66 @@
+
+    /**
+     * Constructor for QgsRelationContext.
+     * \param project associated project
```
```suggestion
```

m-kuhn commented on src/core/qgsrelationcontext.h:
```diff
@@ -0,0 +1,66 @@
+    /**
+     * Get the associated project
+     *
+     * If there is no associated project, then will return the global
```
```suggestion
     * If there is no associated project, then it will return the global
```

m-kuhn commented on src/core/qgsrelationcontext.h:
```diff
@@ -0,0 +1,66 @@
+    ~QgsRelationContext();
+
+    /**
+     * Copies constructor
```
```suggestion
     * Copy constructor
```

m-kuhn commented on src/core/qgsrelationcontext.h:
```diff
@@ -0,0 +1,66 @@
+    QgsRelationContext( const QgsRelationContext &other );
+
+    /**
+     * Copy operator
```
```suggestion
     * Assignment operator
```

m-kuhn commented on src/core/qgsrelationmanager.h:
```diff
@@ -43,6 +43,11 @@ class CORE_EXPORT QgsRelationManager : public QObject
     explicit QgsRelationManager( QgsProject *project = nullptr );
 
+    /**
+     * Get the relation context
```
```suggestion
     * Gets the relation context
```

m-kuhn reviewed (commented):
Good job!

I have left quite a couple of comments, they are pretty much all of them just for a cleaner API documentation in the end and rather quick to get done.

dmarteau commented:
@m-kuhn, changes applied

m-kuhn reviewed (approved):
Thanks :+1:

dmarteau commented:
Pushed again to fix spelling errors.

dmarteau commented:
@m-kuhn Qgis tests passes , but there is travis code layout tests that fails and I can't figure why.

Gustry commented:
@dmarteau Spell check, you should rename `CANT` to something else I guess.

roya0045 commented:
maybe CANNOT will be ok.

Gustry commented:
Well, better to rename to district or similar. It's a French dataset with 'Canton' originally

m-kuhn committed:
Ignore test file in spell check

dmarteau committed:
Add missing parameter documentation

dmarteau commented:
Is this normal that the last check takes more than one day ?

m-kuhn commented:
> Is this normal that the last check takes more than one day ?

Ping @jef-n

nirvn commented:
Warning: I'm pretty sure this has led to QField crashing when opening a feature form that includes relations. I'll confirm within the next 48 hours.

dmarteau commented:
> Warning: I'm pretty sure this has led to QField crashing when opening a feature form that includes relations. I'll confirm within the next 48 hours.

@nirvn It should not, as we have tried to keep the previous behavior. If that's confirmed, please try to reduce the case as much as possible as I have currently no way to test QField build.

nirvn commented:
@dmarteau , @m-kuhn , just built QGIS with this PR reverted, QField crasher is gone.

This PR definitively lead QField to crash when opening a feature form containing relations. That said, it's unclear whether this PR or QField is guilty :)

nirvn commented:
@dmarteau , @m-kuhn , well, I can't duplicate crash today with my QGIS master build sync'ed to the very latest revision. This was seemingly a ghost in the machine.

dmarteau commented:
> @dmarteau , @m-kuhn , well, I can't duplicate crash today (after reverting the git revert) with my QGIS master build sync'ed to the very latest revision. This was seemingly a ghost in the machine.

Glad to here that. Anyway there is a potetential issue if one try to keep alive a relation object past the life of the associated project.

m-kuhn commented:
@nirvn if you get the crash again, could you try to catch a trace?
